uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Service/LeaveRequest.php

```diff
@@ -375,4 +383,46 @@ private function skipBalanceChangeUpdate($params) {
 
     return ($leaveRequestDatesNotChanged || $toilDatesAndToAccrueNotChanged) && !$updateBalanceChange;
   }
+
+  /**
+   * Checks If the current user can cancel a TOIL request type that has past dates.
+   *
+   * @param array $params
+   *
+   * @return bool
+   */
+  private function canCanCancelTOILWithPastDates($params) {
+    return $this->leaveRequestRightsService->canCancelToilWithPastDates($params['contact_id'], $params['type_id']);
+  }
+
+  /**
+   * Checks whether the leave request is of type TOIL and has dates in the
+   * past.
+   *
+   * @param array $params
+   *
+   * @return bool
+   */
+  private function isTOILWithPastDates($params) {
```

Comment:
This looks like duplication of code in the BAO. Can you extract it to another class and use it in both places?

Reply:
good Idea. The best place it can be is in the Leave BAO. It could be a public method used by both classes.